### PR TITLE
Repo Refactor and Crowdfund 

### PR DESCRIPTION
**Pull request by 0xCooki**
- Refactored testing suite to distinguish fuzz and e2e tests
- Refactored revert pattern to use custom errors
- Replaced `bondingCurveActive` boolean with a custom `Phase` enum, which is used to cycle through the various stages of each token in it's life-cycle. 
- Added first draft of Crowdfund prior to bonding curve phase (artist earns 90%, crate 10%)

### CODE REVIEW

File: src/CrateFactoryV1.sol [hunk 1]
```diff
@@ -29,15 +27,12 @@ contract CrateFactoryV1 is Ownable2Step, ReentrancyGuard {
         nonReentrant
         returns (address)
     {
-        address sender = LibMulticaller.sender();
```
Comment (joeyiny) [inline src/CrateFactoryV1.sol:32]: Why remove multicaller?

File: src/CrateFactoryV1.sol [hunk 2]
```diff
@@ -29,15 +27,12 @@ contract CrateFactoryV1 is Ownable2Step, ReentrancyGuard {
...
-
-        require(msg.value == launchCost, "Did not send correct launch cost.");
-        address clone = Clones.cloneDeterministic(tokenImplementation, _saltedSalt(sender, salt));
+        if (msg.value != launchCost) revert InsufficientPayment();
```
Comment (joeyiny) [inline src/CrateFactoryV1.sol]: note: this should be `msg.value < launchCost`

File: src/CrateFactoryV1.sol (hunk 1 above)
Comment (joeyiny) [inline src/CrateFactoryV1.sol:32]: Why remove multicaller?

File: src/CrateFactoryV1.sol [hunk 3]
```diff
@@ -29,15 +27,12 @@ contract CrateFactoryV1 is Ownable2Step, ReentrancyGuard {
...
+        address clone = Clones.cloneDeterministic(tokenImplementation, _saltedSalt(msg.sender, salt));
         CrateTokenV1 newToken = CrateTokenV1(clone);
         allTokens.push(address(newToken));
         emit TokenLaunched(address(newToken), name, symbol);
-        newToken.initialize(uniswapV2Router02, name, symbol, address(this), sender, songURI);
-
+        newToken.initialize(uniswapV2Router02, name, symbol, address(this), msg.sender, songURI);
```
Comment (joeyiny) [inline src/CrateFactoryV1.sol]: `msg.sender` should stay as `sender` and use libmulticaller

File: src/CrateTokenV1.sol [hunk 4]
```diff
@@ -1,41 +1,31 @@
...
+import {IUniswapV2Router02} from "src/interfaces/IUniswapV2RouterV2.sol";
+import {ICrateV1} from "src/interfaces/ICrateV1.sol";
 
 // The total supply is 117,000 tokens
 // Once the bonding curve has sold out 80,000 tokens, the other 37,000 are put in Uniswap with the total ETH in the contract.
 // The LP tokens are then burned, so no one can pull the liquidity.
+contract CrateTokenV1 is ERC20Upgradeable, ReentrancyGuard, ICrateV1 {
+    uint256 private constant MAX_SUPPLY = 117_000e18;
+    uint256 private constant MAX_CURVE_SUPPLY = 80_000e18;
+    uint256 private constant CROWDFUND_THRESHOLD = 60_000e18;
```
Comment (joeyiny) [inline src/CrateTokenV1.sol:16]: I'm assuming `CROWDFUND_THRESHOLD` is only temporarily set to `60_000e18`, this pattern will eventually switch to a USDC goal right?

This is fine for now just confirming

File: src/CrateTokenV1.sol [hunk 5]
```diff
@@ -54,100 +44,123 @@ contract CrateTokenV1 is ERC20Upgradeable, ReentrancyGuard {
...
-        // Take fees out of ETH, then see how many tokens you can buy with the remaining amount.
-        require(bondingCurveActive, "Bonding curve ended");
-        require(tokensInCurve > 0, "Bonding curve is sold out");
+    /// PUBLIC ///
 
+    function buyWithEth(uint256 minTokensReceivable) external payable {
         uint256 preFee =
             (msg.value * (CRATE_FEE_PERCENT + ARTIST_FEE_PERCENT)) / (CRATE_FEE_PERCENT + ARTIST_FEE_PERCENT + 1 ether);
-        uint256 netValue = msg.value - preFee;
+        uint256 netValue = ((msg.value - preFee) * 999) / 1000;
```
Comment (joeyiny) [inline src/CrateTokenV1.sol:61]: what is this part? `* 999) / 1000`

File: src/CrateTokenV1.sol [hunk 6]
```diff
@@ -54,100 +44,123 @@ contract CrateTokenV1 is ERC20Upgradeable, ReentrancyGuard {
...
-        require(tokensToBuy >= minTokensReceivable, "Slippage tolerance exceeded.");
-
-        //If you sent enough eth, then buy tokens.
+        if (tokensToBuy == 0) revert Zero();
+        if (tokensToBuy < minTokensReceivable) revert SlippageToleranceExceeded();
         buy(tokensToBuy);
     }
 
     function buy(uint256 _amount) public payable nonReentrant {
-        address sender = LibMulticaller.sender();
```
Comment (joeyiny) [inline src/CrateTokenV1.sol:82]: should keep Multicaller sender

File: src/CrateTokenV1.sol [hunk 7]
```diff
@@ -54,100 +44,123 @@ contract CrateTokenV1 is ERC20Upgradeable, ReentrancyGuard {
...
-        require(msg.value >= totalPayment, "Not enough Ether to complete purchase.");
-        // Calculate the minimum amount of tokens that should be received based on the slippage tolerance
-        uint256 minTokens = _amount - ((_amount * SLIPPAGE_TOLERANCE) / 10_000);
-
-        require(balanceOf(address(this)) >= minTokens, "Slippage tolerance exceeded.");
+        uint256 totalPayment;
+        uint256 crateFee;
+        uint256 artistFee;
+
+        if (phase == Phase.CROWDFUND) {
```
Comment (joeyiny) [inline src/CrateTokenV1.sol:83]: I am worried about this big code block from line 77 to 101.

There is some repetitive logic, and I am typically not a fan of nested `if` blocks.

Can we refactor this block into cleaner modules/functions/variables, and use the [Return Early pattern](https://medium.com/swlh/return-early-pattern-3d18a41bba8) instead of nested if-elses?

Comment (joeyiny) [reply src/CrateTokenV1.sol:83]: We should also emit an event when transitioning from phases Crowdfund -> bonding curve

File: src/CrateTokenV1.sol [hunk 8]
```diff
@@ -54,100 +44,123 @@ contract CrateTokenV1 is ERC20Upgradeable, ReentrancyGuard {
...
     function sell(uint256 _amount, uint256 minEtherReceivable) external nonReentrant {
-        address sender = LibMulticaller.sender();
+        if (phase != Phase.BONDING_CURVE) revert WrongPhase();
+        if (balanceOf(msg.sender) < _amount) revert InsufficientTokens();
+        if (_amount < 10 ** decimals()) revert MustSellAtLeastOneToken();
+
+        /// @dev cache to use below for transfer
+        uint256 amount = _amount;
+        if (tokensInCurve + _amount >= CROWDFUND_THRESHOLD) {
+            phase = Phase.CROWDFUND;
```
Comment (joeyiny) [inline src/CrateTokenV1.sol]: We should never be able to go back to the Crowdfund phase

File: src/CrateTokenV1.sol [hunk 9]
```diff
@@ -199,31 +210,11 @@ contract CrateTokenV1 is ERC20Upgradeable, ReentrancyGuard {
...
-        require(amountToken > 0 && amountETH > 0 && liquidity > 0, "Liquidity addition failed.");
+        if (amountToken == 0 || amountETH == 0 || liquidity == 0) revert Zero();
         (bool protocolFeePaid,) = protocolFeeDestination.call{value: 0.3 ether}("");
-        require(protocolFeePaid, "Failed to pay protocol fee");
-
+        if (!protocolFeePaid) revert TransferFailed();
         emit LiquidityAdded(amountToken, amountETH, liquidity);
     }
-
-    function _update(address from, address to, uint256 value) internal override {
```
Comment (joeyiny) [inline src/CrateTokenV1.sol:211]: Why was this removed?

**Review by joeyiny — CHANGES_REQUESTED** [review]
Looks good mostly, just had some minor changes. Feel free to ping me to chat about any of them.

File: src/CrateFactoryV1.sol (hunk 1 above)
Comment (0xCooki) [reply src/CrateFactoryV1.sol:32]: What was it achieving? AFAIK the only time it was invoked was to retrieve the sender, which can be done with `msg.sender` or `_msgSender()` via Context.sol if you're worried about other calls from contracts?

File: src/CrateTokenV1.sol (hunk 4 above)
Comment (0xCooki) [reply src/CrateTokenV1.sol:16]: Yes, exactly

File: src/CrateTokenV1.sol (hunk 5 above)
Comment (0xCooki) [reply src/CrateTokenV1.sol:61]: Tests kept failing due to precision loss - at a higher level I'd suggest removing the `buyWithEth` function entirely but I can see the appeal

File: src/CrateTokenV1.sol (hunk 9 above)
Comment (0xCooki) [reply src/CrateTokenV1.sol:211]: achieved nothing

File: src/CrateTokenV1.sol (hunk 8 above)
Comment (0xCooki) [reply src/CrateTokenV1.sol]: How should sales in the bonding curve phase be handled if they'd push the tokens held by the contract back into the crowd fund phase?

File: src/CrateTokenV1.sol (hunk 9 above)
Comment (0xCooki) [reply src/CrateTokenV1.sol:211]: Seemed unnecessary given checks in other functions, am happy to add re-add though

File: src/CrateFactoryV1.sol (hunk 1 above)
Comment (joeyiny) [reply src/CrateFactoryV1.sol:32]: iirc it was so that a user can launch a token and buy some in one transaction
